### PR TITLE
VACMS-9921: Expand Cypress accessibility test coverage.

### LINKED ISSUE #9921: Expand Cypress-Axe accessibility test coverage
## Description
We should test a broader range of pages to be more confident about our accessibility.  Foreground/background text contrast in particular should be found fairly easily, but we have a number of issues with that.

## Acceptance Criteria
- [ ] Cypress accessibility tests have been expanded to cover other view modes for all content types.
- [ ] Other pages (content views, etc) are covered by accessibility tests.  Pretty much anything accessible by a `content_admin` user and used in the course of normal work.

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

Closes #9921.

Adds accessibility testing for all node forms and a view for a published node of each type where there is an existing published node.

I figure this is a good start.

Theoretically this should be safe to merge, but it's obviously not if we have existing accessibility test failures on these pages.

### CODE REVIEW

**Comment by ndouglas** [comment]
This is failing because they completely rewrote the config forms.  I'll open an issue to address.

**Comment by ndouglas** [comment]
Hey @laflannery, I'm working on surfacing these Cypress errors better.  Is there any additional information you'd like in these comments?

**Comment by laflannery** [comment]
A few questions - these are all based on my axe plugin so there's a chance I'm making some wild assumptions here:

1. When I run the browser extension axe is able to flag why there might be a contrast issue.  For example:

image
or
image

Is any information like this coming over for contrast issues? It is very often useful to know if this is flagged for a clear violation or something you are going to have to test manually

2. Would we be able to get the full html for the element? I also find this useful, if there are elements with the same class it could help to find the issue easier. It looks like we have the Element Location (Target in Failures above) but could we get Element Source?
image

3. Deque also provides in their plugin output links to their docs on how you might solve the issues. Could these be added? And if so could they be useful for developers that are perhaps not as familiar with accessibility and the errors they are seeing?

**Comment by ndouglas** [comment]
I'll see what I can pull it -- I don't think I have that much information available, but it's possible that I can install a plugin or something like that and get some additional information that'd be helpful.

**Comment by ndouglas** [comment]
@laflannery I'm not seeing any additional information I can pull in re the reasoning that a check might be failing, but hopefully the HTML, help page, and other information is useful 🙂

**Comment by ndouglas** [comment]
brace yourself, I'm gonna run a test on all of the new pages I added... let's see if the script craps out, and where it craps out 🤔

**Comment by laflannery** [comment]
@ndouglas I think this other info will be very useful, the help link especially for non-obvious issues.  This is great, thank you!  Also...I'm bracing for impact...

**Comment by laflannery** [comment]
@ndouglas I believe you can exclude certain elements from the tests if you don't want to report on them.  You can exclude the below (they are 3rd party widgets that we cannot control):

1. `.leaflet-marker-icon`
**Will resolve this:**

> /node/add/vet_center_cap
> ID: image-alt
> Impact: critical
> Tags: cat.text-alternatives, wcag2a, wcag111, section508, section508.22.a, ACT
> Description: Ensures  elements have alternate text or a role of none or presentation
> Help: [Images must have alternate text](https://dequeuniversity.com/rules/axe/4.4/image-alt?application=axeAPI)
> Nodes:
> 
> 
> HTML: 
> Impact: critical
> Target: .leaflet-marker-icon
> Summary: Fix any of the following:
> Element does not have an alt attribute
> aria-label attribute does not exist or is empty
> aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
> Element has no title attribute
> Element's default semantics were not overridden with role="none" or role="presentation"

2. `#jsd-widget`
**Will resolve this:**

> /help/vamc/about-locations-content-for-vamcs/about-alerts-and-operating-statuses
> ID: frame-title
> Impact: serious
> Tags: cat.text-alternatives, wcag2a, wcag241, wcag412, section508, section508.22.i
> Description: Ensures <iframe> and <frame> elements have an accessible name
> Help: [Frames must have an accessible name](https://dequeuniversity.com/rules/axe/4.4/frame-title?application=axeAPI)
> Nodes:
> 
> HTML: <iframe scrolling="no" name="JSD widget" id="jsd-widget" style="height: 10px; width: 11px; position: fixed; border: 0px; bottom: 0px; right: 0px; z-index: 2147483647; transition: height 0s ease 0s;"></iframe>
> Impact: serious
> Target: #jsd-widget
> Summary: Fix any of the following:
> Element has no title attribute
> aria-label attribute does not exist or is empty
> aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
> Element's default semantics were not overridden with role="none" or role="presentation"

**Comment by ndouglas** [comment]
@laflannery Sure can, we're already excluding the 8000+-item select widget 🙂  Should be a pretty quick change to make.

**Comment by EWashb** [comment]
@laflannery do you know if this is ok to close?

**Comment by laflannery** [comment]
@EWashb The issue with this is that when we opened this up to so many more pages, it found so many more issues.  I believe (if I'm understanding this correctly), these existing failures would block other work from going out and therefore we need to deal with these before cypress-axe can be officially considered "expanded".  I am tracking these - we already resolved some and only have about 10 left, some simpler, some more complex.  I would be ok with temporarily excluding the complex errors if we felt we wanted to get this completed sooner and those complex issues were the only blockers. Let me know if you wanted to chat about this and I can explain further

**Comment by laflannery** [comment]
- Some new instances of existing issues but no new issues which is excellent.
- No instances of previous issues we already resolved (progress!!)
- @ndouglas The 2 elements from last week we were going to exclude are still being reported on.  Was that just not done yet or is that not working?

**Comment by ndouglas** [comment]
👀  ah, I bet I gotta do `img.leaflet-marker-icon` and `iframe#jsd-widget`.  Didn't have my frontend-dev brain turned on.

File: tests/cypress/integration/accessibility/accessibility.spec.js [hunk 1]
```diff
@@ -44,7 +181,12 @@ const axeContext = {
   include: [["body"]],
   exclude: [
     [
-      "#edit-menu-menu-parent", // 8700-item select elements apparently break accessibility tests
+      // 8700-item select elements apparently break accessibility tests.
+      "#edit-menu-menu-parent",
+      // Not our widget, not our problem.
+      ".leaflet-marker-icon",
```
Comment (ndouglas) [inline tests/cypress/integration/accessibility/accessibility.spec.js:187]:
```suggestion
      "img.leaflet-marker-icon",
```

File: tests/cypress/integration/accessibility/accessibility.spec.js [hunk 2]
```diff
@@ -44,7 +181,12 @@ const axeContext = {
...
+      // Not our widget, not our problem.
+      "#jsd-widget",
```
Comment (ndouglas) [inline tests/cypress/integration/accessibility/accessibility.spec.js:189]:
```suggestion
      "iframe#jsd-widget",
```

**Comment by ndouglas** [comment]
There we go @laflannery.

File: tests/cypress/integration/accessibility/accessibility.spec.js [hunk 3]
```diff
@@ -3,21 +3,158 @@
...
+  "/resources",
+  // Leadership Listing
+  "/node/add/leadership_listing",
+  "/dayton-health-care/about-us/leadership",
+  // NCA Facility
+  "/node/add/nca_facility",
+  "/nca-facilities/locations/fargo-national-cemetery",
+  // News Release
+  "/node/add/press_release",
+  "/orlando-health-care/news-releases/orlando-va-health-care-system-hosts-global-war-on-terrorism-wall-sept-6-15",
```
Comment (swirtSJW) [inline tests/cypress/integration/accessibility/accessibility.spec.js:54]: I am a bit concerned that we are targetting so  many different vha health care system paths  (I count at least 17).  Targeting specific content is brittle enough.  Health care system names/paths rarely change but there is still an X percent chance, and by targetting 17 different systems we now have a 17X chance.  Our paths don't have different templates per system, so is there another reason to target different health care systesm?

Comment (ndouglas) [reply tests/cypress/integration/accessibility/accessibility.spec.js:54]: Are you saying I should pick e.g. `central-ohio-health-care` and have a news release, office, staff profile, etc from that specific system?

I don't have any objection to that, and don't have a reason for picking the specific nodes I did, other than that at the time I originally opened this PR, those were relatively recent examples of each node that were published, had meaningful content, etc.  I can go back and restrict my choices to a single system.

Alternatively, we could switch this to address non-pre-existing-content pages (like the `/node/add` forms) and roll the accessibility tests for read views into #11493.  That should be less brittle and probably an all-around better approach longterm.

Thoughts?

Comment (swirtSJW) [reply tests/cypress/integration/accessibility/accessibility.spec.js:54]: Sorry @ndouglas somehow I missed this comment.
Yes I was saying pick a single system like `central-ohio-health-care` and go with it.  That would make it less prone to failure than smei-randomly checking a smattering of systems.

I like your idea of rolling them all into /node/add/* that would solves my concern a different way and seems to be even less brittle... though probably tougher to do.

Comment (ndouglas) [reply tests/cypress/integration/accessibility/accessibility.spec.js:54]: Suppose I could roll this whole thing into #11493 🤔

**Comment by laflannery** [comment]
FYI Edmund is working on the color contrast issues now which will clear up 75% of these (#11397 & #11455)

**Comment by laflannery** [comment]
These remaining issues are things that don't seem super straight forward to me (i.e. not just changing colors) - my thought is that after we handle Drupal 10 maybe I can get with Edmund to see what he thinks about these.

**Comment by ndouglas** [comment]
@laflannery Sounds great!  I'm delighted there are so _few_ issues!

**Comment by ndouglas** [comment]
Superseded by #12620.

**Pull request closed without merging**